portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -26,40 +32,43 @@
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
 
-  private List<String> messages;
   private static final String COMMENT = "comment-input";
   private static final String NAME = "name-input";
-
-  @Override
-  public void init() {
-    messages = new ArrayList<>();
-  }
+  private DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+  private static final int maxResults = 10;
 
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
     // Construct comment from user request.
     String fullComment = request.getParameter(NAME) + ": " + request.getParameter(COMMENT);
-    messages.add(fullComment);
 
     response.setContentType("text/html;");
     response.getWriter().println(fullComment);
 
+    Entity commEntity = new Entity("Comment");
+    commEntity.setProperty("content", fullComment);
+    datastore.put(commEntity);
+
     // Redirect back to main page.
     response.sendRedirect("/index.html");
   }
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    Query query = new Query("Comment");
+    PreparedQuery results = datastore.prepare(query);
+    List<Entity> messages = results.asList(FetchOptions.Builder.withLimit(maxResults));
+
     String json = convertToJson(messages);
     response.setContentType("application/json;");
     response.getWriter().println(json);
   }
 
-  private String convertToJson(List<String> messageList) {
+  private String convertToJson(List<Entity> messages) {
     String json = "{\"history\": [";
-    for (int i = 0; i < messageList.size(); i++) {
-      json += "\"" + messages.get(i) + "\"";
-      if (i != messageList.size() - 1) {
+    for (int i = 0; i < messages.size(); i++) {
+      json += "\"" + (String) messages.get(i).getProperty("content") + "\"";
```

Comment:
if you store the name and comment separately, your json could look like this:

[{name: "Lily", comment: "Lily's comment #1"}, {name: "Megan", comment: "Megan's comment!"}]

Also same comment about using a constant here instead of the string "content".